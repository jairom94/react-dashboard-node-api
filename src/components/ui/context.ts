import { createContext, useContext } from "react";

export const RadioGroupContext = createContext<{
    sale?:boolean;
    value?:string;
    name:string;
    onChange:(nameSelected:string)=>void;
}>({
    // sale:false,
    name:'',
    value:'',
    onChange:()=>{},
})

export function useRadioGroup() {
    const radioGroupValue = useContext(RadioGroupContext)
    return radioGroupValue
}

export const CheckGroupContext = createContext<{
    name:string;
    selectsItem:string[];
    onSelect:(value:string)=>void;
    onUnSelect:(value:string)=>void;
    clearSelects:()=>void;    
}>({
    name:'',
    selectsItem:[],
    onSelect:()=>{},
    onUnSelect:()=>{},
    clearSelects:()=>{},    
})

export function useCheckGroup() {
    const checkGroupValue = useContext(CheckGroupContext)
    return checkGroupValue
}