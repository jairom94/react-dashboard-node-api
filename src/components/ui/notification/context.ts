import { createContext, useContext } from "react";
import type { Notification } from "./types";

interface NotificationContextType{
    addNoti:(notification:Notification)=>void;
}

export const NotificationContext = createContext<NotificationContextType | null>(null)

export function useNotification(){
    const context = useContext(NotificationContext)
    if(!context){
        throw new Error('useNotification must be used within NotificationProvider')        
    }
    return context
}