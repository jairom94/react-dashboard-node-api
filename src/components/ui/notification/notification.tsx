import { useEffect, useState } from "react";
import type { Icons } from "./types";


interface NotificationProps {
    message: string; 
    type: Icons; 
    onDelete:()=>void; 
    createdAt:number;  
}
const Notification = ({ message, type, onDelete,createdAt }: NotificationProps) => { 
  const [progress, setProgress] = useState(0);
  const duration = 5000
  useEffect(() => {
    const intervalMs = 10;

    const interval = setInterval(() => {
      const elapsed = Date.now() - createdAt;
      const percentage = Math.min((elapsed / duration) * 100,100);
      setProgress(percentage);

      if (elapsed >= duration) {
        clearInterval(interval);
        onDelete();
      }
    }, intervalMs);

    return () => clearInterval(interval);
  }, [createdAt, onDelete]); 
  
  const icons = {
  success: "M8.5 11.5 11 14l4-4m6 2a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  info: "M10 11h2v5m-2 0h4m-2.592-8.5h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  warning: "M12 13V8m0 8h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  error: "m15 9-6 6m0-6 6 6m6-3a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
}; 
  return (
    <li className={`notification-item ${type}`}>        
      <div className="notification-content">
        <div className="notification-icon">
          <svg aria-hidden="true" fill="none" viewBox="0 0 24 24">
            <path
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d={icons[type]}
            />
          </svg>
        </div>
        <div className="notification-text">{message}</div>
      </div>
      <button 
      onClick={onDelete}
      className="notification-icon notification-close">
        <svg aria-hidden="true" fill="none" viewBox="0 0 24 24">
          <path
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M6 18 17.94 6M18 18 6.06 6"
          ></path>
        </svg>
      </button>
      <div
        className="notification-progress-bar"
        style={{ width: `${progress}%` }}
      />
    </li>
  );
};

export default Notification;
