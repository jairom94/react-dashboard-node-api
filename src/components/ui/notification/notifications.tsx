import { createPortal } from 'react-dom';
import NotificationComponent from './notification';
import type { Notification } from './types';


interface NotficationsProps {
    notifications:Notification[];
    handleNotifications:(notifications:Notification[])=>void;
}
const Notfications = ({notifications,handleNotifications}:NotficationsProps) => {
    const portalNotifications = document.querySelector('#notifications-root')
    function handleClickDelete(notificationId:string){
        const deleteNotis = notifications.filter(n => n.id !== notificationId)
        handleNotifications(deleteNotis)
    }
    return createPortal(
    <ul className='notification-container'>
        { notifications.map(notification => (
            <NotificationComponent 
            key={notification.id} 
            onDelete={()=>handleClickDelete(notification.id)} 
            message={notification.message} 
            createdAt={notification.createdAt}
            type={notification.type} 
            />
        )) }
    </ul>
    ,portalNotifications as HTMLDivElement)
};

export default Notfications;
