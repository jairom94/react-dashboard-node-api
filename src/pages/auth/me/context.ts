import { createContext, useContext } from "react";
import type { UserResponse } from "../types";

export const UserContext = createContext<{
    user:UserResponse | null;
    onUserLogged:()=>void;
    onUserLogout:()=>void;
}>({
  user:null,
  onUserLogged: ()=>{},
  onUserLogout: ()=>{},
});

export function useUserInformation() {
  const user = useContext(UserContext);
  return user;
}
