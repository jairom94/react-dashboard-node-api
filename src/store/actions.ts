import { AxiosError } from "axios";
import type { AppThunk } from ".";
import { createAdvert, deleteAdvert, detailAdvert, getAdverts, getTags } from "../pages/adverts/service";
import type { Advert, Tag } from "../pages/adverts/types";
import type { Modal } from "../pages/adverts/partials/types";

//LOADED ADVERTS
type AdvertsLoadedPending = {
    type:'adverts/loaded/pending',
    payload:null
}
type AdvertsLoadedRejected = {
    type:'adverts/loaded/rejected';
    payload:Error;
}
type AdvertsLoadedFullFilled = {
    type:'adverts/loaded/fullfilled',
    payload:Advert[]
}

//DELETE ADVERT
type AdvertsDeletedPending = {
    type:'adverts/deleted/pending',
    payload:null
}
type AdvertsDeletedRejected = {
    type:'adverts/deleted/rejected';
    payload:Error;
}
type AdvertsDeletedFullFilled = {
    type:'adverts/deleted/fullfilled',
    payload:Advert;
}

//CREATE ADVERT
type AdvertsCreatedPending = {
    type:'adverts/created/pending',
    payload:null
}
type AdvertsCreatedRejected = {
    type:'adverts/created/rejected';
    payload:Error;
}
type AdvertsCreatedFullFilled = {
    type:'adverts/created/fullfilled',
    payload:Advert;
}

//LOADED TAGS
type TagsLoadedPending = {
    type:'tags/loaded/pending',
    payload:null
}
type TagsLoadedRejected = {
    type:'tags/loaded/rejected';
    payload:Error;
}
type TagsLoadedFullFilled = {
    type:'tags/loaded/fullfilled',
    payload:Tag[]
}

//MODAL
type ModalShowFullFilled = {
    type:'modal/show/fullfilled',
    payload:Modal<Advert>;
}

type ModalCloseFullFilled = {
    type:'modal/close/fullfilled',
    // payload:Modal<Advert>;
}

//MODAL
export const modalShowFullFilled = (modal:Modal<Advert>):ModalShowFullFilled => {
    // console.log(modal);    
    return {
        type:"modal/show/fullfilled",
        payload:modal
    }
}
export const modalCloseFullFilled = ():ModalCloseFullFilled=> {
    // console.log(modal);    
    return {
        type:"modal/close/fullfilled",
        // payload:modal
    }
}





//LOADED TAGS
export const advertsLoadedPending = ():AdvertsLoadedPending => ({
    type: "adverts/loaded/pending",
    payload:null
})
export const advertsLoadedRejected = (error:Error):AdvertsLoadedRejected => ({
    type:"adverts/loaded/rejected",
    payload:error
})
export const advertsLoadedFullFilled = (adverts:Advert[]):AdvertsLoadedFullFilled => ({
    type:"adverts/loaded/fullfilled",
    payload:adverts
})

//DELETE ADVERT
export const advertsDeletedPending = ():AdvertsDeletedPending => ({
    type: "adverts/deleted/pending",
    payload:null
})
export const advertsDeletedRejected = (error:Error):AdvertsDeletedRejected => ({
    type:"adverts/deleted/rejected",
    payload:error
})
export const advertsDeletedFullFilled = (advertDeleted:Advert):AdvertsDeletedFullFilled => ({
    type:"adverts/deleted/fullfilled",
    payload:advertDeleted
})

//CREATED ADVERT
export const advertsCreatedPending = ():AdvertsCreatedPending => ({
    type: "adverts/created/pending",
    payload:null
})
export const advertsCreatedRejected = (error:Error):AdvertsCreatedRejected => ({
    type:"adverts/created/rejected",
    payload:error
})
export const advertsCreatedFullFilled = (advert:Advert):AdvertsCreatedFullFilled => ({
    type:"adverts/created/fullfilled",
    payload:advert
})
type UiResetError = {
  type: "ui/reset-error";
};

//LOADED TAGS
export const tagsLoadedPending = ():TagsLoadedPending => ({
    type: "tags/loaded/pending",
    payload:null
})
export const tagsLoadedRejected = (error:Error):TagsLoadedRejected => ({
    type:"tags/loaded/rejected",
    payload:error
})
export const tagsLoadedFullFilled = (tags:Tag[]):TagsLoadedFullFilled => ({
    type:"tags/loaded/fullfilled",
    payload:tags
})

//LOADED TAGS ACTION
export function tagsLoaded():AppThunk<Promise<void>>{
    return async function (dispatch) {
        dispatch(tagsLoadedPending())
        try {
            const tags = await getTags()
            dispatch(tagsLoadedFullFilled(tags))
        } catch (error) {
            if(error instanceof AxiosError){
                dispatch(tagsLoadedRejected(error))
            }
            throw error
        }
    }
}



//LOADED ADVERTS ACTION
export function advertsLoaded(filterByTag:string):AppThunk<Promise<void>>{
    return async function (dispatch) {
        dispatch(advertsLoadedPending())
        try {
            const adverts = await getAdverts(filterByTag)
            dispatch(advertsLoadedFullFilled(adverts))
        } catch (error) {
            if(error instanceof AxiosError){
                dispatch(advertsLoadedRejected(error))
            }
            throw error
        }
    }
}

//CREATED ADVERT ACTION
export function advertsCreated(newAdvert:Advert):AppThunk<Promise<Advert>>{
    return async function (dispatch) {
        dispatch(advertsCreatedPending())
        try {
            const { id } = await createAdvert(newAdvert)
            const advert = await detailAdvert((id ?? '').toString())
            dispatch(advertsCreatedFullFilled(advert))
            return advert
        } catch (error) {
            if(error instanceof AxiosError){
                dispatch(advertsCreatedRejected(error))
            }
            throw error
        }
    }
}

//DELETED ADVERT ACTION
export function advertsDeletedOne(advertId:string):AppThunk<Promise<void>>{
    return async function (dispatch) {
        dispatch(advertsDeletedPending())
        try {
            const advert = await detailAdvert(advertId)                      
            await deleteAdvert(advertId)
            dispatch(advertsDeletedFullFilled(advert))
        } catch (error) {
            if(error instanceof AxiosError){
                dispatch(advertsDeletedRejected(error))
            }
            throw error
        }
    }
}

export const uiResetError = ():UiResetError => ({
    type:"ui/reset-error"
})

export type Actions = 
| AdvertsLoadedPending
| AdvertsLoadedRejected
| AdvertsLoadedFullFilled
| AdvertsCreatedPending
| AdvertsCreatedRejected
| AdvertsCreatedFullFilled
| AdvertsDeletedPending
| AdvertsDeletedRejected
| AdvertsDeletedFullFilled
| TagsLoadedPending
| TagsLoadedRejected
| TagsLoadedFullFilled
| ModalShowFullFilled
| ModalCloseFullFilled
| UiResetError