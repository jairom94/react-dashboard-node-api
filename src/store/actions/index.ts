import type { AdvertsActions } from "./adverts";
import type { AuthActions } from "./auth";
import type { ModalActions } from "./modal";
import type { SessionActions } from "./session";
import type { TagsActions } from "./tags";
import type { UiActions } from "./ui";

export * from './auth';
export * from './session';
export * from './adverts';
export * from './modal';
export * from './ui';
export * from './tags';



export type Actions = 
| AuthActions
| SessionActions
| AdvertsActions
| UiActions
| TagsActions
| ModalActions