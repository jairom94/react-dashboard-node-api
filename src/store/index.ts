import { composeWithDevTools } from "@redux-devtools/extension";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { useDispatch, useSelector } from "react-redux";
import * as thunk from 'redux-thunk'
import type { Actions } from "./actions";
import * as reducers from './reducer'


const rootReducer = combineReducers(reducers)

export default function configureStore(preloadedState:Partial<reducers.State>){
    const store = createStore(
        rootReducer,
        preloadedState as never,
    composeWithDevTools(
        applyMiddleware(
            thunk.withExtraArgument(),
        )
    )
    )

    return store
}

export type AppStore = ReturnType<typeof configureStore>
export type AppGetState = AppStore["getState"];
export type RootState = ReturnType<AppGetState>;
export type AppDispatch = AppStore["dispatch"];

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
// useSelector conoce la forma del estado
export const useAppSelector = useSelector.withTypes<RootState>();

export type AppThunk<ReturnType = void> = thunk.ThunkAction<
  ReturnType,
  RootState,
  undefined,
  Actions
>;