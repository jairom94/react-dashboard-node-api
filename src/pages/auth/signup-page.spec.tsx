import { render,screen } from "@testing-library/react";
import SignPage from "./signup-page";
import { Router } from "react-router";
import { NotificationContext } from "../../components/ui/notification/context";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";

import { authSignup } from "../../store/actions"
vi.mock("../../store/actions")
describe("SignupPage", () => {
  const state = {
    auth: false,
    session: "",
    adverts: {
      loaded: false,
      data: [],
    },
    tags: [],
    filters: {
      name: "",
      tags: [],
      price: [0, 0],
      range: [0, 0],
    },
    ui: {
      pending: false,
      error: null,
    },
    modal: {
      data: null,
      type: "",
      visible: false,
    },
  };
  const renderComponent = () =>
    render(
      <Provider
        store={{
          getState: () => state,
          // @ts-expect-error: we only need to mock the methods we use
          subscribe: () => {},
          // @ts-expect-error: we only need to mock the methods we use
          dispatch: () => {},
        }}
      >
        <NotificationContext.Provider value={{ addNoti: vi.fn() }}>
          <Router
            navigator={{
              createHref: () => "",
              go: () => {},
              push: () => {},
              replace: () => {},
            }}
            location={{ pathname: "" }}
          >
            <SignPage />
          </Router>
        </NotificationContext.Provider>
      </Provider>,
    );
  test("should render correctly", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch authSignup",async ()=>{
    renderComponent();
    // const name = screen.getByLabelText("Nombres");
    // const username = screen.getByLabelText("Nombre de usuario");
    // const email = screen.getByLabelText("E-mail");
    // const password = screen.getByLabelText("Contrase√±a");
    const submit = screen.getByRole("button", { name: /enviar/i });

    await userEvent.click(submit)
    expect(authSignup).toHaveBeenCalledWith({
      name: "",
      username: "",
      email: "",
      password: "",
    });
  })
});
