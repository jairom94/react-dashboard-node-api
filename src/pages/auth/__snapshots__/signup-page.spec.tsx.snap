// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignupPage > should render correctly 1`] = `
<div>
  <div
    class="bg-gray-800 min-h-dvh flex flex-col justify-center items-center"
  >
    <div
      class="bg-white px-5 py-8 rounded-lg md:container-md container flex flex-col justify-center gap-4"
    >
      <h3
        class="text-2xl font-medium text-gray-800"
      >
        Formulario de registro
      </h3>
      <form
        class="flex flex-col gap-2"
      >
        <div
          class="form-field flex flex-col gap-1"
        >
          <label>
            Nombres
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            name="name"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field flex flex-col gap-1"
        >
          <label>
            Nombre de usuario
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field flex flex-col gap-1"
        >
          <label>
            E-mail
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            name="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="form-field flex flex-col gap-1"
        >
          <label>
            Contraseña
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-field flex flex-col"
        >
          <input
            class="rounded-md border px-4 py-2 bg-emerald-600 hover:bg-emerald-500 transition-colors duration-300 cursor-pointer disabled:opacity-50 disabled:pointer-events-none"
            name="enviar"
            type="submit"
          />
        </div>
      </form>
      <div
        class="flex justify-center"
      >
        <p
          class="font-light text-gray-700"
        >
          ¿Ya tienes una cuenta?
           
          <a
            class="text-md font-medium text-emerald-600 underline"
            data-discover="true"
            href=""
          >
            Iniciar sesión
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
